src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -738,12 +913,174 @@ export class SavedObjectsRepository {
       id,
       type,
       updated_at: time,
-      version: encodeHitVersion(response),
+      version: encodeHitVersion(updateResponse),
+      ...(this._registry.isMultiNamespace(type) && {
+        namespaces: updateResponse.get._source.namespaces,
+      }),
       references,
       attributes,
     };
   }
 
+  async addNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsAddNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
+
+    const { version, namespace, refresh = DEFAULT_REFRESH_SETTING } = options;
+
+    const rawId = this._serializer.generateRawId(undefined, type, id);
+    const preflightResult = await this.preflightCheckIncludesNamespace(type, id, namespace);
+    const existingNamespaces = getSavedObjectNamespaces(undefined, preflightResult);
+
+    if (existingNamespaces?.length) {
+      // there should never be a case where a multi-namespace object does not have any existing namespaces
+      // however, it is a possibility if someone manually modifies the document in Elasticsearch
+      const intersection = existingNamespaces.filter(x => namespaces.includes(x));
+      if (intersection.length) {
+        throw SavedObjectsErrorHelpers.createBadRequestError(
+          `${rawId} already exists in the following namespace(s): ${intersection.join(', ')}`
+        );
```

Comment:
I'm starting on Phase 1.5 and thinking through the UX of sharing / un-sharing...

I realized the Bad Request error here may prove to be challenging in the next phases.
For example, if Dashboard "Foo" has a reference to Visualization "Bar", and they both exist in Space A...
* I share "Bar" to Space B because I want to use it in a different dashboard there
* Later, I want to share "Foo" to Space B too

In this scenario, if we use a "dumb approach" for sharing, the operation for "Bar" will fail, and we have to parse the error / ignore it in the UX. Alternatively, if we try to use a "smart approach" we will have to get all of the reference saved objects, and avoid adding namespaces which would cause conflict. Either way this results in some messy logic on the frontend.

A similar scenario can happen with `deleteNamespaces` for namespaces that don't exist on a given saved object -- that currently also throws a Bad Request error.

I wonder if we would be better off changing these functions here and allowing these cases to succeed. That would make things simpler all around. I can't think of a great reason to continue throwing a Bad Request like we are doing here.

Thoughts?

Reply:
I'm not opposed to simplifying this and letting these requests succeed

Reply:
Done in fa5f361.